include/gridtools/fn/backend2/gpu.hpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * GridTools
+ *
+ * Copyright (c) 2014-2021, ETH Zurich
+ * All rights reserved.
+ *
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#pragma once
+
+#include "../../common/cuda_util.hpp"
+#include "../../common/hymap.hpp"
+#include "../../meta.hpp"
+#include "../../sid/allocator.hpp"
+#include "../../sid/concept.hpp"
+#include "../../sid/contiguous.hpp"
+#include "../../sid/multi_shift.hpp"
+
+namespace gridtools::fn::backend {
+    namespace gpu_impl_ {
+        template <class BlockSizes>
+        struct gpu {
+            using block_sizes_t = BlockSizes;
+        };
+
+        template <class BlockSizes, class Dim>
+        using block_size_at_dim = meta::second<meta::mp_find<BlockSizes, Dim>>;
+
+        template <class BlockSizes, class Sizes>
+        GT_FUNCTION_DEVICE auto global_thread_index() {
+            using keys_t = get_keys<Sizes>;
+            using ndims_t = meta::length<keys_t>;
+            static_assert(ndims_t::value > 0);
+            if constexpr (ndims_t::value == 1) {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using values_t = typename keys_t::template values<int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x);
+            } else if constexpr (ndims_t::value == 2) {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using block_dim_y = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 1>>;
+                using values_t = typename keys_t::template values<int, int>;
+                return values_t(
+                    blockIdx.x * block_dim_x::value + threadIdx.x, blockIdx.y * block_dim_y::value + threadIdx.y);
+            } else {
+                using block_dim_x = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 0>>;
+                using block_dim_y = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 1>>;
+                using block_dim_z = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 2>>;
+                using values_t = typename keys_t::template values<int, int, int>;
+                return values_t(blockIdx.x * block_dim_x::value + threadIdx.x,
+                    blockIdx.y * block_dim_y::value + threadIdx.y,
+                    blockIdx.z * block_dim_z::value + threadIdx.z);
+            }
+        }
+
+        template <class Index, class Sizes>
+        GT_FUNCTION_DEVICE bool in_domain(Index const &index, Sizes const &sizes) {
+            auto in_bounds = tuple_util::device::transform(std::less(), index, sizes);
+            return tuple_util::device::fold(std::logical_and(), wstd::move(in_bounds));
```

Comment:
there is `tuple_uti::all_of`: 
```c++
 return tuple_util::device::all_off(std::less(), index, sizes);
```
should work.

Reply:
Or even:
```c++
constexpr GT_DEVICE in_domain = tuple_util::device::all_of(std::less());
```

Reply:
GPU compilers did not like the second option, so I went with the first one.